File: go/parse_component.go
package vbml

func ParseComponent(defaultHeight, defaultWidth int, props map[string]any, component Component) (Board, error) {
	if component.RawCharacters != nil {
		return component.RawCharacters, nil
	}

	width, height := componentDimensions(defaultHeight, defaultWidth, component.Style)
	if component.RandomColors != nil {
		return randomColors(height, width, component.RandomColors.Colors), nil
	}

	template := component.Template
	justify := JustifyLeft
	align := AlignTop
	if component.Style != nil {
		if component.Style.Justify != "" {
			justify = component.Style.Justify
		}
		if component.Style.Align != "" {
			align = component.Style.Align
		}
	}

	text := emojisToCharacterCodes(template)
	text = parseProps(props, text)
	text = SanitizeSpecialCharacters(text)
	words := splitWords(width, text)
	lines := getLinesFromWords(width, words)

	codes := make(Board, 0, len(lines))
	for _, line := range lines {
		row, err := convertCharactersToCharacterCodes(line)
		if err != nil {
			return nil, err
		}
		codes = append(codes, row)
	}

	codes = verticalAlign(height, align, codes)
	codes = horizontalAlign(width, justify, codes)
	return renderComponent(createEmptyBoard(height, width), codes), nil
}

func ParseAbsoluteComponent(defaultHeight, defaultWidth int, props map[string]any, component Component) (AbsoluteComponent, error) {
	characters, err := ParseComponent(defaultHeight, defaultWidth, props, component)
	if err != nil {
		return AbsoluteComponent{}, err
	}

	absoluteComponent := AbsoluteComponent{
		Characters: characters,
		X:          0,
		Y:          0,
	}
	if component.Style != nil && component.Style.AbsolutePosition != nil {
		absoluteComponent.X = component.Style.AbsolutePosition.X
		absoluteComponent.Y = component.Style.AbsolutePosition.Y
	}
	return absoluteComponent, nil
}

func componentDimensions(defaultHeight, defaultWidth int, style *ComponentStyle) (width, height int) {
	width = defaultWidth
	height = defaultHeight
	if style == nil {
		return width, height
	}
	if style.Width != 0 {
		width = style.Width
	}
	if style.Height != 0 {
		height = style.Height
	}
	return width, height
}

File: python/pyvbml/parse_component.py
"""Parse component.

Port of Vestaboard/vbml/src/parseComponent.ts
"""

from __future__ import annotations

from .character_codes import COLOR_CODES, convert_characters_to_character_codes
from .create_empty_board import create_empty_board
from .horizontal_align import horizontal_align
from .random_colors import random_colors
from .render_component import render_component
from .template_parts import resolve_template_lines
from .types import Align, IVBMLComponent, Justify, VBMLProps
from .vertical_align import vertical_align


def parse_component(
    default_height: int,
    default_width: int,
    props: VBMLProps,
    component: IVBMLComponent,
) -> list[list[int]]:
    """Parse component."""
    if "rawCharacters" in component:
        return component["rawCharacters"]

    style = component.get("style") or {}
    width = style.get("width") or default_width
    height = style.get("height") or default_height

    if "randomColors" in component:
        colors = component["randomColors"].get("colors") or COLOR_CODES
        return random_colors(height, width, colors)

    empty_component = create_empty_board(height, width)
    template = component.get("template", "") if "template" in component else ""
    justify = style.get("justify") or Justify.LEFT
    align = style.get("align") or Align.TOP

    lines = resolve_template_lines(width, props, template)
    codes = [convert_characters_to_character_codes(line) for line in lines]
    codes = vertical_align(height, align, codes)
    codes = horizontal_align(width, justify, codes)
    return render_component(empty_component, codes)


def parse_absolute_component(
    default_height: int,
    default_width: int,
    props: VBMLProps,
    component: IVBMLComponent,
) -> dict:
    """Parse absolute component."""
    style = component.get("style") or {}
    abs_pos = style.get("absolutePosition")
    return {
        "characters": parse_component(default_height, default_width, props, component),
        "x": abs_pos.get("x", 0) if abs_pos else 0,
        "y": abs_pos.get("y", 0) if abs_pos else 0,
    }
